### PR TITLE
Use common dispatcher for public client outbound #2

### PR DESCRIPTION
**Pull request by vytautas-karpavicius**
**What changed?**
Same as #4523, but with fixed middleware pointer, which required the revert #4534.

authMiddleware was stored as a pointer, which is not equal to nil which is checked when wrapping a middleware here: https://github.com/yarpc/yarpc-go/blob/dev/api/middleware/outbound.go#L59

Also updated unit test to ensure true nil is used when auth is not configured.

---

Stop using `DispatcherProvider` for creating public client. Instead use common dispatcher from rpc.Factory.

Builder for public client outbound was created which also includes auth middleware initialization. It was previously located within server.go itself. Moving it here allows consolidating all logic related logic in one place and covering it with unit tests.

**Why?**
To eventually delete `DispatcherProvider` entirely.

**How did you test it?**
Existing tests and new unit tests.

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/rpc/outbounds.go [hunk 1]
```diff
@@ -21,12 +21,57 @@
...
+		authMiddleware = &authOutboundMiddleware{authProvider}
+	}
+
+	return publicClientOutbound{config.PublicClient.HostPort, authMiddleware}, nil
+}
+
+type publicClientOutbound struct {
+	address        string
+	authMiddleware middleware.UnaryOutbound
+}
```
Comment (mantas-sidlauskas) [inline common/rpc/outbounds.go:68]: move this up

**Review by mantas-sidlauskas — APPROVED** [review]

File: common/rpc/outbounds.go (hunk 1 above)
Comment (vytautas-karpavicius) [reply common/rpc/outbounds.go]: done
